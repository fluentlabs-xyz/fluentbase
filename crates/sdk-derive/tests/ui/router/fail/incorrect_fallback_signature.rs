#![allow(dead_code)]
extern crate alloc;
extern crate fluentbase_sdk;
use fluentbase_sdk::{
    basic_entrypoint,
    derive::{router, Contract},
    SharedAPI,
};

#[derive(Contract)]
struct App<SDK> {
    sdk: SDK,
}

pub trait RouterAPI {
    // Fallback must have no parameters and no return value
    fn fallback(&self, data: u64) -> bool;

    // Regular method for the router to have something to process
    fn regular_method(&self);
}

#[router(mode = "solidity")]
impl<SDK: SharedAPI> RouterAPI for App<SDK> {
    fn fallback(&self, data: u64) -> bool {
        // Should cause an error as fallback cannot have parameters or return value
        data > 10
    }

    #[function_id("regularMethod()", validate(true))]
    fn regular_method(&self) {
        // Normal method
    }
}

impl<SDK: SharedAPI> App<SDK> {
    pub fn deploy(&self) {}
}

basic_entrypoint!(App);
