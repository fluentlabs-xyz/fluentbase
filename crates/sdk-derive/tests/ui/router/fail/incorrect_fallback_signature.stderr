error: Fallback method must have signature 'fn fallback(&self)' with no parameters and no return value
  --> tests/ui/router/fail/incorrect_fallback_signature.rs:25:5
   |
25 |     fn fallback(&self, data: u64) -> bool {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Failed to process router implementation due to method parsing errors
  --> tests/ui/router/fail/incorrect_fallback_signature.rs:23:1
   |
23 | #[router(mode = "solidity")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `router` (in Nightly builds, run with -Z macro-backtrace for more info)
